### ISSUE TITLE
prePuller.hook.pullOnlyOnChanges not pulling only on changes

### ISSUE THREAD

**Issue by yuvipanda** [jupyterhub/zero-to-jupyterhub-k8s#2172]
### Bug description

I've `prePuller.hook.pullOnlyOnChanges` (added in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2066) set to true, but the hook prePuller is still being triggered for each deploy.

### How to reproduce
Relevant config is:

```yaml
jupyterhub:
  prePuller:
    hook:
      pullOnlyOnChanges: true
```

Running helm upgrade with `--debug`, I get:

```
history.go:56: [debug] getting history for release r-staging
upgrade.go:123: [debug] preparing upgrade for r-staging
upgrade.go:131: [debug] performing update for r-staging
upgrade.go:303: [debug] creating upgraded release for r-staging
client.go:282: [debug] Starting delete for "r-staging-default-priority" PriorityClass
client.go:122: [debug] creating 1 resource(s)
client.go:282: [debug] Starting delete for "r-staging-user-placeholder-priority" PriorityClass
client.go:122: [debug] creating 1 resource(s)
client.go:282: [debug] Starting delete for "hook-image-puller" DaemonSet
client.go:311: [debug] daemonsets.apps "hook-image-puller" not found
client.go:122: [debug] creating 1 resource(s)
client.go:282: [debug] Starting delete for "hook-image-awaiter" ServiceAccount
client.go:311: [debug] serviceaccounts "hook-image-awaiter" not found
client.go:122: [debug] creating 1 resource(s)
client.go:282: [debug] Starting delete for "hook-image-awaiter" Role
client.go:311: [debug] roles.rbac.authorization.k8s.io "hook-image-awaiter" not found
client.go:122: [debug] creating 1 resource(s)
client.go:282: [debug] Starting delete for "hook-image-awaiter" RoleBinding
client.go:311: [debug] rolebindings.rbac.authorization.k8s.io "hook-image-awaiter" not found
client.go:122: [debug] creating 1 resource(s)
client.go:282: [debug] Starting delete for "hook-image-awaiter" Job
client.go:311: [debug] jobs.batch "hook-image-awaiter" not found
client.go:122: [debug] creating 1 resource(s)
client.go:491: [debug] Watching for changes to Job hook-image-awaiter with timeout of 5m0s
client.go:519: [debug] Add/Modify event for hook-image-awaiter: ADDED
client.go:558: [debug] hook-image-awaiter: Jobs active: 1, jobs failed: 0, jobs succeeded: 0
client.go:519: [debug] Add/Modify event for hook-image-awaiter: MODIFIED
client.go:282: [debug] Starting delete for "hook-image-puller" DaemonSet
client.go:282: [debug] Starting delete for "hook-image-awaiter" ServiceAccount
client.go:282: [debug] Starting delete for "hook-image-awaiter" Role
client.go:282: [debug] Starting delete for "hook-image-awaiter" RoleBinding
client.go:282: [debug] Starting delete for "hook-image-awaiter" Job
client.go:201: [debug] checking 27 resources for changes
client.go:464: [debug] Looks like there are no changes for PodDisruptionBudget "user-placeholder"
client.go:464: [debug] Looks like there are no changes for PodDisruptionBudget "user-scheduler"
client.go:464: [debug] Looks like there are no changes for ServiceAccount "hub"
...
```

This just repeats, even with no changes at all.

### Versions

helm version: v3.5.4
kubernetes version: v1.18.16-gke.502
z2jh version:  0.11.1-n438.h7e6a66e9 (master at time of bug report)

**Comment by consideRatio** [comment]
@yuvipanda ah thanks for the bug report.

To debug this is a bit trickier than debugging other features, because debugging this means accessing the k8s api-server using the `lookup` function - something that doesn't happen with `helm template` I think.

The way the feature works is by either rendering the `hook-image-puller` DaemonSet and `hook-image-awaiter` Job resource conditionally to the truthyness of the _named template_ named `jupyterhub.imagePuller.daemonset.hook.install` (in `_helpers-daemonset.tpl`).

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/b804bac3ad6acf8b656f6108a5a18b95d9d22c10/jupyterhub/templates/image-puller/_helpers-daemonset.tpl#L200-L221

Perhaps if we add a `#` comment evaluating that named function in some k8s resource, allowing you to see what it evaluated to, then we could understand what goes on better?

**Comment by consideRatio** [comment]
I've debugged this and drawn the conclusion that:

1. Labels of the DaemonSet resource include the chart name, which makes any chart version change lead to a checksum change.
2. The labels are generated depending the context from which the named template is called, and hence the checksum.

The fix for both is to not checksum too much to ensure we don't let the checksum vary based on labels of no relevance.

**Comment by consideRatio** [comment]
Merged to try it in a live JupyterHub and could verify it to work! This is made available as part of `0.11.1-n451.h198fc854` and later.
